### PR TITLE
Tpwdcee 6259 feature

### PR DESCRIPTION
**Pull request by tomazpu**

### CODE REVIEW

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Coverage increased per file
===========================
+ Gateway/Request/CreditCardTransactionFactory.php  4
         

Coverage decreased per file
===========================
- Gateway/Command/WirecardCommand.php  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5187253&bid=17019061)

File: view/frontend/web/js/view/payment/method-renderer/vault.js [hunk 1]
```diff
@@ -139,6 +140,65 @@ define([
...
+                }
+            });
+        },
+        appendFormData: function (data, form) {
+            for (let key in data) {
+                if (key !== "form-url" && key !== "form-method") {
+                    form.append($("<input />", {
+                        type: "hidden",
+                        name: key,
+                        value: data[key]
```
Comment (rinnhofer) [inline view/frontend/web/js/view/payment/method-renderer/vault.js]: Codacy Issue found: [Generic Object Injection Sink](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5187253)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Coverage increased per file
===========================
+ Gateway/Request/CreditCardTransactionFactory.php  4
         

Coverage decreased per file
===========================
- Gateway/Command/WirecardCommand.php  -1
         

Clones added
============
- view/frontend/web/js/view/payment/method-renderer/vault.js  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5187253&bid=17019061)

File: view/frontend/web/js/view/payment/method-renderer/vault.js [hunk 2]
```diff
@@ -139,6 +140,65 @@ define([
...
+         * Handle 3Ds credit card transactions within callback
+         * @param response
+         */
+        redirectCreditCard: function (response) {
+            let result = {};
+            result.data = {};
+            let appendFormData = this.appendFormData.bind(this);
+            $.ajax({
+                url: url.build("wirecard_elasticengine/frontend/callback"),
+                dataType: 'json',
```
Comment (rinnhofer) [inline view/frontend/web/js/view/payment/method-renderer/vault.js]: Codacy Issue found: [Strings must use doublequote.](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5187253)

File: view/frontend/web/js/view/payment/method-renderer/vault.js [hunk 3]
```diff
@@ -139,6 +140,65 @@ define([
...
+                        let form = $("<form />", {
+                            action: result.data["form-url"],
+                            method: result.data["form-method"]
+                        });
+                        appendFormData(result.data, form);
+                        form.appendTo("body").submit();
+                    }
+                },
+                error: function (err) {
+                    this.addErrorMessageAndRedirect(errorList);
```
Comment (rinnhofer) [inline view/frontend/web/js/view/payment/method-renderer/vault.js]: Codacy Issue found: ['errorList' is not defined.](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5187253)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Coverage increased per file
===========================
+ Gateway/Request/CreditCardTransactionFactory.php  4
         

Coverage decreased per file
===========================
- Gateway/Command/WirecardCommand.php  -1
         

Clones added
============
- view/frontend/web/js/view/payment/method-renderer/vault.js  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5187253&bid=17019061)

File: view/frontend/web/js/view/payment/method-renderer/vault.js [hunk 4]
```diff
@@ -139,6 +140,65 @@ define([
...
+                }
+            });
+        },
+        appendFormData: function (data, form) {
+            for (let key in data) {
+                if (key !== "form-url" && key !== "form-method") {
+                    form.append($("<input />", {
+                        type: "hidden",
+                        name: key,
+                        value: data[key]
```
Comment (rinnhofer) [inline view/frontend/web/js/view/payment/method-renderer/vault.js]: Codacy Issue found: [Generic Object Injection Sink](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5187253)

Comment (tomazpu) [reply view/frontend/web/js/view/payment/method-renderer/vault.js]: WIll be handled in the javascript refactoring.

File: Gateway/Command/WirecardCommand.php [hunk 5]
```diff
@@ -88,6 +87,9 @@ public function __construct(
     public function execute(array $commandSubject)
     {
         $transaction = $this->transactionFactory->create($commandSubject);
+        if ($transaction instanceof CreditCardTransaction) {
```
Comment (ozada) [inline Gateway/Command/WirecardCommand.php]: Couldn't you get type information from transaction object itself? or transfer this logic in transaction inside?

File: view/frontend/web/js/view/payment/method-renderer/vault.js [hunk 6]
```diff
@@ -139,6 +140,65 @@ define([
...
+                    window.location.replace(data["redirect-url"]);
+                });
+            }
+        },
+        /**
+         * Handle 3Ds credit card transactions within callback
+         * @param response
+         */
+        redirectCreditCard: function (response) {
+            let result = {};
```
Comment (ozada) [inline view/frontend/web/js/view/payment/method-renderer/vault.js]: you could declare dict with empty keys and avoid line 167.

File: view/frontend/web/js/view/payment/method-renderer/vault.js [hunk 7]
```diff
@@ -139,6 +140,65 @@ define([
             this.seamlessResponse = response;
             this.placeOrder();
         },
+        afterPlaceOrder: function () {
+            if (this.seamlessResponse.hasOwnProperty("acs_url")) {
+                this.redirectCreditCard(this.seamlessResponse);
+            } else {
+                // Handle redirect for Non-3D transactions
+                $.ajax({
+                    url: url.build("wirecard_elasticengine/frontend/redirect"),
```
Comment (ozada) [inline view/frontend/web/js/view/payment/method-renderer/vault.js]: You could declare urls as constants at he top in settings.

Comment (tomazpu) [reply view/frontend/web/js/view/payment/method-renderer/vault.js]: as the whole file has URLs as this I will leave it but add it to the refactoring US in the sprint.

File: Gateway/Command/WirecardCommand.php (hunk 5 above)
Comment (tomazpu) [reply Gateway/Command/WirecardCommand.php]: No, as this functionality is not built into the paymentSDK it is not possible.

File: view/frontend/web/js/view/payment/method-renderer/vault.js (hunk 7 above)
Comment (ozada) [reply view/frontend/web/js/view/payment/method-renderer/vault.js]: OK

File: view/frontend/web/js/view/payment/method-renderer/vault.js [hunk 8]
```diff
@@ -139,6 +140,66 @@ define([
...
+                }
+            });
+        },
+        appendFormData: function (data, form) {
+            for (let key in data) {
+                if (key !== "form-url" && key !== "form-method") {
+                    form.append($("<input />", {
+                        type: "hidden",
+                        name: key,
+                        value: data[key]
```
Comment (rinnhofer) [inline view/frontend/web/js/view/payment/method-renderer/vault.js]: Codacy Issue found: [Generic Object Injection Sink](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5187253)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Coverage increased per file
===========================
+ Gateway/Request/CreditCardTransactionFactory.php  4
         

Coverage decreased per file
===========================
- Gateway/Command/WirecardCommand.php  -1
         

Clones added
============
- view/frontend/web/js/view/payment/method-renderer/vault.js  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/magento2-ee/pullRequest?prid=5187253&bid=17019061)

**Pull request merged**